plugins/catalog-backend/src/ingestion/processors/microsoftGraph/read.ts

```diff
@@ -0,0 +1,285 @@
+import { buildOrgHierarchy } from '../util/org';
+import { MicrosoftGraphClient } from './client';
+import {
+  MICROSOFT_GRAPH_GROUP_ID_ANNOTATION,
+  MICROSOFT_GRAPH_TENANT_ID_ANNOTATION,
+  MICROSOFT_GRAPH_USER_ID_ANNOTATION,
+} from './constants';
+
+export function normalizeEntityName(name: string): string {
+  return name
+    .trim()
+    .toLocaleLowerCase()
+    .replace(/[^a-zA-Z0-9_\-\.]/g, '_');
+}
+
+export async function readMicrosoftGraphUsers(
+  client: MicrosoftGraphClient,
+  options?: { userFilter?: string },
+): Promise<{
+  users: UserEntity[]; // With all relations empty
+}> {
+  const entities: UserEntity[] = [];
+  const picturePromises: Promise<void>[] = [];
+
+  for await (const user of client.getUsers({
+    filter: options?.userFilter,
+    select: ['id', 'displayName', 'mail', ''],
+  })) {
+    if (!user.id || !user.displayName || !user.mail) {
+      continue;
+    }
+
+    const name = normalizeEntityName(user.mail);
+    const entity: UserEntity = {
+      apiVersion: 'backstage.io/v1alpha1',
+      kind: 'User',
+      metadata: {
+        name,
+        annotations: {
+          [MICROSOFT_GRAPH_USER_ID_ANNOTATION]: user.id!,
+        },
+      },
+      spec: {
+        profile: {
+          displayName: user.displayName || undefined,
+          email: user.mail || undefined,
+
+          // TODO: Additional fields?
+          // jobTitle: user.jobTitle || undefined,
+          // officeLocation: user.officeLocation || undefined,
+          // mobilePhone: user.mobilePhone || undefined,
+        },
+        memberOf: [],
+      },
+    };
+
+    // Download the photos in parallel, otherwise it can take quite some time
+    const loadPhoto = async () => {
+      entity.spec.profile!.picture = await client.getUserPhotoWithSizeLimit(
+        user.id!,
+        // We are limiting the photo size, as users with full resolution photos
+        // can make the Backstage API slow
+        120,
+      );
+    };
+
+    picturePromises.push(loadPhoto());
+    entities.push(entity);
+  }
+
+  // Wait for all photos to be downloaded
+  await Promise.all(picturePromises);
+
+  return { users: entities };
+}
+
+export async function readMicrosoftGraphOrganizations(
+  client: MicrosoftGraphClient,
+): Promise<{
+  rootGroup?: GroupEntity; // With all relations empty
+}> {
+  const organizations = await client.getOrganization();
+
+  if (organizations) {
+    for (const entry of organizations) {
+      if (!entry.id || !entry.displayName) {
+        continue;
+      }
+
+      // For now we expect a single root orgranization
+      const name = normalizeEntityName(entry.displayName);
+      const rootGroup: GroupEntity = {
+        apiVersion: 'backstage.io/v1alpha1',
+        kind: 'Group',
+        metadata: {
+          name: name,
+          description: entry.displayName,
+          annotations: {
+            [MICROSOFT_GRAPH_TENANT_ID_ANNOTATION]: entry.id,
```

Comment:
Not tenant right

Reply:
No it's correct here. This is not a group but the tenant that is used as the root group.

Reply:
Ah yeah I saw further down in the code, should have deleted this